### PR TITLE
Automl

### PR DESCRIPTION
**Pull request by JordanHolland**
Base code for AutoML.

Autogluon for AutoML training and testing.

Takes care of #1 using Sklearn train_test_split.

Takes care of #2: binary and multiclass ROC, PR, and confusion matrix. Further, by using the leaderboard built into AutoGluon we have more statistics on inference time, training time, and overall scores.

### CODE REVIEW

File: src/nprintml/automl.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import autogluon as ag
+from autogluon import TabularPrediction as task
+from sklearn.model_selection import train_test_split
+from sklearn.preprocessing import LabelBinarizer
+from sklearn.metrics import roc_curve, auc, precision_recall_curve, \
+        average_precision_score, confusion_matrix, ConfusionMatrixDisplay
+
+import seaborn as sns
+sns.set_style("ticks")
+sns.set_context(rc={"lines.linewidth": 1.75})
```
Comment (JordanHolland) [inline src/nprintml/automl.py:15]: I'm of the opinion that we set defaults and provide the most basic use case graphs with reasonable defaults (i.e. we've used these defaults in papers) and if user's want more intense graphs we leave the models around for them to use.

File: src/nprintml/automl.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        self.outpath = outpath
+        self.qd = {
+          0: 'medium_quality_faster_train',
+          1: 'good_quality_faster_inference_only_refit',
+          2: 'high_quality_fast_inference_only_refit',
+          3: 'best_quality',
+          }
+
+    def run(self, test_size=.30, eval_metric='accuracy', quality=0,
+            time_limits=5*60, n_threads=1):
```
Comment (JordanHolland) [inline src/nprintml/automl.py:30]: We're going to need cli flags for

- test size: 0.00 - 1.00 (size of test split)
- eval_metric: metric to optimize for, [available list:](https://auto.gluon.ai/stable/api/autogluon.task.html#tabular)
- quality: [available list:](https://auto.gluon.ai/stable/api/autogluon.task.html#tabular)
- time_limits: max time limit in seconds for each model
- n_threads: number of threads for training

File: src/nprintml/automl.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+from sklearn.preprocessing import LabelBinarizer
+from sklearn.metrics import roc_curve, auc, precision_recall_curve, \
+        average_precision_score, confusion_matrix, ConfusionMatrixDisplay
+
+import seaborn as sns
+sns.set_style("ticks")
+sns.set_context(rc={"lines.linewidth": 1.75})
+
+
+class AutoML:
```
Comment (JordanHolland) [inline src/nprintml/automl.py:18]: There may be an argument for removing graph code from the AutoML class, but we've already written about 90% of the AutoML code we will likely write for nPrintML.

File: src/nprintml/automl.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                             nthreads_per_trial=n_threads)
+
+        return predictor
+
+    def test(self, predictor, test_data):
+        leaderboard = predictor.leaderboard(test_data, silent=True)
+        leaderboard = leaderboard.set_index('model').sort_index()
+        leaderboard.to_csv(self.outpath / 'leaderboard.scv')
+
+    def generate_graphs(self, predictor, test_data):
```
Comment (JordanHolland) [inline src/nprintml/automl.py:53]: Tested with binary and multiclass problems, which I've pushed (don't really care where they end up, just gives us toy data for classification problems)

**Comment by JordanHolland** [comment]
You'll also notice I pushed two csv files under the test directory. I used these to test the AutoML framework for both multiclass and binary classification problems. They are probably good to have around, though you may have better intuition where they belong.

File: src/nprintml/automl.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        fpr, tpr, thresholds = roc_curve(y_true, y_proba)
+        cl_auc = auc(fpr, tpr)
+        label = None
+        if cl:
+            label = '{0} - AUC: {1:.2f}'.format(cl, cl_auc)
+        else:
+            label = 'AUC: {0:.2f}'.format(cl_auc)
+        ax = sns.lineplot(x=fpr, y=tpr, label=label)
+
+        return ax
```
Comment (jesteria) [inline src/nprintml/automl.py:130]:
```suggestion
        return sns.lineplot(x=fpr, y=tpr, label=label)
```

File: src/nprintml/automl.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+        self.finalize_graph(self.outpath / 'roc.pdf', x_lim=[0.0, 1.0],
+                            y_lim=[0.0, 1.05], x_label='False Positive Rate',
+                            y_label='True Positive Rate',
+                            legend_loc='lower right')
+
+    def _make_binary_roc(self, y_true, y_proba, cl=None):
+        fpr, tpr, thresholds = roc_curve(y_true, y_proba)
+        cl_auc = auc(fpr, tpr)
+        label = None
```
Comment (jesteria) [inline src/nprintml/automl.py:123]: (superfluous)

```suggestion
```

File: src/nprintml/automl.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        if len(classes) == 2:
+            ax = self._make_binary_roc(y_true_bin.ravel(), y_proba.ravel())
+        else:
+            for i, cl in enumerate(classes):
+                ax = self._make_binary_roc(y_true_bin[:, i], y_proba[:, i], cl=cl)
+
+        # Split up the line styles to make them unique
+        linestyles = ['-', '--', '-.', ':']
+        for i, line in enumerate(ax.lines):
+            line.set_linestyle(linestyles[i % len(linestyles)])
```
Comment (jesteria) [inline src/nprintml/automl.py:113]: The `mod` way works great so you might prefer it -- just as an alternative --

```suggestion
        linestyles = ('-', '--', '-.', ':')
        for (linestyle, line) in zip(itertools.cycle(linestyles),
                                     ax.lines):
            line.set_linestyle(linestyle)
```

File: src/nprintml/automl.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        precision, recall, thresholds = precision_recall_curve(y_true, y_proba)
+        cl_ap = average_precision_score(y_true, y_proba)
+        label = None
+        if cl:
+            label = '{0} - AP: {1:.2f}'.format(cl, cl_ap)
+        else:
+            label = 'AP: {0:.2f}'.format(cl_ap)
+        ax = sns.lineplot(x=recall, y=precision, label=label)
+
+        return ax
```
Comment (jesteria) [inline src/nprintml/automl.py:100]:
```suggestion
        return sns.lineplot(x=recall, y=precision, label=label)
```

File: src/nprintml/automl.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            line.set_linestyle(linestyles[i % len(linestyles)])
+
+        self.finalize_graph(self.outpath / 'pr.pdf', x_lim=[0.0, 1.0],
+                            y_lim=[0.0, 1.05], legend_loc='lower left',
+                            x_label='Recall', y_label='Precision')
+
+    def _make_binary_pr(self, y_true, y_proba, cl=None):
+        precision, recall, thresholds = precision_recall_curve(y_true, y_proba)
+        cl_ap = average_precision_score(y_true, y_proba)
+        label = None
```
Comment (jesteria) [inline src/nprintml/automl.py:93]:
```suggestion
```

File: src/nprintml/automl.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        if len(classes) == 2:
+            ax = self._make_binary_pr(y_true_bin.ravel(), y_proba.ravel())
+        else:
+            for i, cl in enumerate(classes):
+                ax = self._make_binary_pr(y_true_bin[:, i], y_proba[:, i], cl=cl)
+
+        # Split up the line styles to make them unique
+        linestyles = ['-', '--', '-.', ':']
+        for i, line in enumerate(ax.lines):
+            line.set_linestyle(linestyles[i % len(linestyles)])
```
Comment (jesteria) [inline src/nprintml/automl.py:84]:
```suggestion
        linestyles = ('-', '--', '-.', ':')
        for (linestyle, line) in zip(itertools.cycle(linestyles),
                                     ax.lines):
            line.set_linestyle(linestyle)
```

**Review by jesteria — COMMENTED** [review]
This all looks _very_ cool :+1: 

Some suggestions -- take 'em or leave 'em :slightly_smiling_face: 

(And, FYI, I use the Github "suggestion" feature in part because it makes it easy to incorporate such changes -- assuming you're happy with them as is.)

**Comment by JordanHolland** [comment]
committed suggested changes, linted and added comments, tested on binary and multiclass, merging.
